### PR TITLE
Chargeback: updated to v2 table api which returns display columns and units

### PR DESCRIPTION
**Pull request by dtaylor113**
Using units to format dates and sort numeric columns.

image

ToDo:
- [ ] When ready, need to test to see if new code works with 4.2 OLM Metering installation
      Merged:  [PR: Update OLM/Marketplace 4.2 package](https://github.com/operator-framework/operator-metering/pull/868) 
     @chancez "...4.2 stuff comes out with 4.2. this relates to RCM/ART publishing our stuff, which they arent doing for 4.2 yet.  ill be working on making it so you can semi-test using 4.2 via OLM, but the manual install should be similar enough to do anything you need also"

JIRA: Chargeback Reports: Use v2 report results API to show field units and which should be shown/hidden - https://jira.coreos.com/browse/CONSOLE-1445

cc @chancez

### CODE REVIEW

**Comment by tlwu2013** [comment]
I know this is still WIP, but just noticed from the screenshot, the table entries are not "`vertical-align: top`" as other tables in Console. Could you help check on that? Thanks!

**Comment by dtaylor113** [comment]
> I know this is still WIP, but just noticed from the screenshot, the table entries are not "`vertical-align: top`" as other tables in Console. Could you help check on that? Thanks!

Nice catch Tony :-)  The Usage table is not virtualized, so I think I need to add another override for non-virtualized tables, or maybe broaden this rule?  Shouldn't all cells be `v.align: top`, regardless of table type?  cc @priley86 
```
.pf-c-table.pf-c-virtualized tr>td {
    vertical-align: top;
}
```

**Comment by priley86** [comment]
We should be able to safely remove these overrides now in `_overrides.scss`:

```
// override .co-m-table-grid [class*='col-'], .co-m-table-grid .row
.pf-c-table.pf-c-virtualized tr > td {
  vertical-align: top;
}

.pf-c-table.pf-c-virtualized tr.vertical-align-middle > td {
  vertical-align: middle;
}

.pf-c-table.pf-c-virtualized tr.vertical-align-middle {
  .co-clusterserviceversion-logo__icon,
  .co-clusterserviceversion-logo {
    align-items: center;
  }
}
```

PF4 is setting `vertical-align: baseline` by default which should work everywhere.

**Comment by priley86** [comment]
I think we have to override this style for now:
```
.co-m-table-grid [class*='col-'], .co-m-table-grid .row {
    vertical-align: middle;
```

Otherwise, things will look out of alignment on some views:
Screen Shot 2019-08-05 at 1 50 30 PM


PF4 uses `vertical-align: baseline`, but `vertical-align: top` can also work here.

**Comment by dtaylor113** [comment]
demo:  https://console-openshift-console.apps.taylordev114.devcluster.openshift.com
Username: kubeadm, Pwd: gdpMA-zkbtB-LtKJI-Lh2yy
28 Chargeback Reports under ns="openshift-metering"

**Comment by chancez** [comment]
Sometimes dates are showing as a relative delta between now and the date rather than just showing the date (period end should be a date not "5 months from now"):
Screen Shot 2019-08-07 at 1 50 49 PM

**Comment by dtaylor113** [comment]
> Sometimes dates are showing as a relative delta between now and the date rather than just showing the date (period end should be a date not "5 months from now"):

Hi Chance, that's the standard datetime display we use in the console UI.  Consistent with all other UI pages.

**Comment by spadgett** [comment]
> > Sometimes dates are showing as a relative delta between now and the date rather than just showing the date (period end should be a date not "5 months from now"):
> 
> Hi Chance, that's the standard datetime display we use in the console UI. Consistent with all other UI pages.

It's a bug in the standard date component. We're trying to show dates within the last 10 minutes or so as relative time, but we're not expecting dates in the future. @dtaylor113 can you open a bug?

**Comment by chancez** [comment]
I think for a reporting system it almost never makes sense to show relative times in the results. It may make sense in the details of the report, but i don't think it does in the report results.

File: frontend/public/components/factory/table.tsx [hunk 1]
```diff
@@ -304,7 +307,7 @@ export const Table = connect<TablePropsFromState,TablePropsFromDispatch,TablePro
       staticFilters: PropTypes.array,
       virtualize: PropTypes.bool,
       currentSortField: PropTypes.string,
-      currentSortFunc: PropTypes.string,
+      currentSortFunc: PropTypes.string || PropTypes.func,
```
Comment (spadgett) [inline frontend/public/components/factory/table.tsx]: You're overloading `sortFunc` to mean completely different things depending on the type that's passed. As mentioned in the previous PR, this should be a separate prop specfically to transform the value for sorting.

File: frontend/public/style/_overrides.scss [hunk 2]
```diff
@@ -387,7 +387,7 @@ kbd {
 }
 
 // override .co-m-table-grid [class*='col-'], .co-m-table-grid .row
-.pf-c-table.pf-c-virtualized tr > td {
+.pf-c-table.pf-c-virtualized tr > td, .co-m-table-grid .pf-c-table [class*='col-'] {
```
Comment (spadgett) [inline frontend/public/style/_overrides.scss]: can you explain the need for this change? i thought `co-m-table-grid` was deprecated and we weren't using it for chargeback

File: frontend/public/components/factory/table.tsx (hunk 1 above)
Comment (dtaylor113) [reply frontend/public/components/factory/table.tsx]: idk, I feel it's appropriate.  When it's a string, it's an index into an array of sortFuncs, when it's a func, it's a direct sortFunc as defined by the app dev.   If I add another property, it seems more confusing to add a `sortConversion` which is the same as `sortFunc`.

Comment (spadgett) [reply frontend/public/components/factory/table.tsx]: >  If I add another property, it seems more confusing to add a sortConversion which is the same as sortFunc.

But it's not the same as `sortFunc`! They're fundamentally different. `sortFunc` gets the value out of the object. `sortConversion` transforms the `sortField` value.

Today you never pass both `sortField` and `sortFunc` together. You've created a special case where if you pass *both* `sortField` and `sortFunc` AND `sortFunc` is a function, it does something special. That's a confusing and unintuitive API. The only way you would know this is to read the code closely (and the code is already a little hard to understand).

Honestly, I feel like we're overthinking this. Why not just pass an optional `sortAsNumber` prop to tell table to treat the column as a number. That's what we need here. If we ever need something more general, we can revisit.

File: frontend/public/components/chargeback.tsx [hunk 3]
```diff
@@ -159,12 +159,22 @@ class ReportsDetails extends React.Component<ReportsDetailsProps> {
...
+  'memory_bytes',
+  'memory_byte_seconds',
+  'seconds']);
+
+const DataCell = ({name, value, unit}) => {
+  if (numericUnits.has(unit)) {
     return <div className="text-right">{_.round(value, 2).toLocaleString()}</div>;
   }
+  if (unit === 'date' || unit === 'time') {
+    return <Timestamp timestamp={value} />;
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: Ok, will be handled via: https://bugzilla.redhat.com/show_bug.cgi?id=1739225

File: frontend/public/components/factory/table.tsx (hunk 1 above)
Comment (dtaylor113) [reply frontend/public/components/factory/table.tsx]: ok, it doesn't seem all that confusing to me :-)  if you specify sortField, that field will be sorted as a string, if you pass sortFunc:string, you are defining a named sortFunction (meaning the function knows what field to sort on, it's in the func name,ex: "getClusterOperatorStatus"), if you define a sortField and sortFunc:func, your sortFunc will be called passing the sortField.

`sortAsNumber` seems like a good solution, I'll implement that -thanks

File: frontend/public/components/chargeback.tsx [hunk 4]
```diff
@@ -159,12 +159,22 @@ class ReportsDetails extends React.Component<ReportsDetailsProps> {
...
-const DataCell = ({name, value}) => {
-  if (_.isFinite(value)) {
+const numericUnits = new Set([
+  'bytes',
+  'byte_seconds',
+  'cpu_core_seconds',
+  'cpu_cores',
+  'memory_bytes',
+  'memory_byte_seconds',
+  'seconds']);
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: fixed

File: frontend/public/components/chargeback.tsx [hunk 5]
```diff
@@ -173,21 +183,32 @@ const DataCell = ({name, value}) => {
   return value;
 };
 
-const DataTable = ({cols, rows}:DataTableProps) => {
+const DataTable = ({cols, rows, schema}:DataTableProps) => {
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: fixed

File: frontend/public/components/chargeback.tsx [hunk 6]
```diff
@@ -173,21 +183,32 @@ const DataCell = ({name, value}) => {
...
   const size = _.clamp(Math.floor(12 / _.size(rows[0])), 1, 4);
   const className = `col-md-${size}`;
+  const toNumber = (value) => _.toNumber(value);
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: using `sortAsNumber`

File: frontend/public/components/chargeback.tsx [hunk 7]
```diff
@@ -159,12 +159,22 @@ class ReportsDetails extends React.Component<ReportsDetailsProps> {
...
+const numericUnits = new Set([
+  'bytes',
+  'byte_seconds',
+  'cpu_core_seconds',
+  'cpu_cores',
+  'memory_bytes',
+  'memory_byte_seconds',
+  'seconds']);
+
+const DataCell = ({name, value, unit}) => {
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: fixed

File: frontend/public/components/chargeback.tsx [hunk 8]
```diff
@@ -173,21 +183,32 @@ const DataCell = ({name, value}) => {
...
 };
 
-const DataTable = ({cols, rows}:DataTableProps) => {
+const DataTable = ({cols, rows, schema}:DataTableProps) => {
   const size = _.clamp(Math.floor(12 / _.size(rows[0])), 1, 4);
   const className = `col-md-${size}`;
+  const toNumber = (value) => _.toNumber(value);
+
+  const getUnit = (col) => {
+    const colSchema:any = _.find(schema.values, {'name': col});
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: fixed

File: frontend/public/components/chargeback.tsx [hunk 9]
```diff
@@ -173,21 +183,32 @@ const DataCell = ({name, value}) => {
   return value;
 };
 
-const DataTable = ({cols, rows}:DataTableProps) => {
+const DataTable = ({cols, rows, schema}:DataTableProps) => {
   const size = _.clamp(Math.floor(12 / _.size(rows[0])), 1, 4);
   const className = `col-md-${size}`;
+  const toNumber = (value) => _.toNumber(value);
+
+  const getUnit = (col) => {
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: fixed

File: frontend/public/components/chargeback.tsx [hunk 10]
```diff
@@ -173,21 +183,32 @@ const DataCell = ({name, value}) => {
...
+    const colSchema:any = _.find(schema.values, {'name': col});
+    return _.get(colSchema, 'unit');
+  };
 
   const DataTableHeader = () => _.map(cols, col => {
+    let sortFn = {};
+    if (numericUnits.has(getUnit(col))) {
+      sortFn = { sortFunc: toNumber };
+    }
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: using `sortAsNumber`

File: frontend/public/components/chargeback.tsx [hunk 11]
```diff
@@ -173,21 +183,32 @@ const DataCell = ({name, value}) => {
...
       sortField: col,
       title: col.replace(/_/g, ' '),
       transforms: [sortable],
       props: { className },
     });
   });
 
-  const DataTableRows = ({componentProps: {data}}) => _.map(data, (r) => _.map(r, (v, k) => {
-    return {title: <DataCell name={k} value={v} />};
+  const DataTableRows = ({componentProps: {data}}) => _.map(data, (r) => _.map(r, (v, c) => {
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: fixed

File: frontend/public/components/chargeback.tsx [hunk 12]
```diff
@@ -243,11 +264,11 @@ class ReportData extends React.Component<ReportDataProps, ReportDataState> {
   }
 
   makeTable(data=this.state.data) {
-    if (_.isEmpty(data)) {
+    if (_.isEmpty(data) || _.isEmpty(data.results)) {
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: fixed

File: frontend/public/components/chargeback.tsx [hunk 13]
```diff
@@ -256,30 +277,37 @@ class ReportData extends React.Component<ReportDataProps, ReportDataState> {
...
 
-  getColumns(data=[]) {
-    const cols = _.keys(data[0]).filter(k => {
-      if (COLS_BLACK_LIST.has(k)) {
-        return false;
-      }
-      return true;
-    });
-    return cols;
+  getColumns(data) {
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: fixed

File: frontend/public/components/chargeback.tsx [hunk 14]
```diff
@@ -256,30 +277,37 @@ class ReportData extends React.Component<ReportDataProps, ReportDataState> {
...
+    const firstRow:any = _.head(data.results);
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: fixed

File: frontend/public/components/chargeback.tsx [hunk 15]
```diff
@@ -256,30 +277,37 @@ class ReportData extends React.Component<ReportDataProps, ReportDataState> {
...
+    return _.sortBy(_.map(firstRow.values, item => item.name));
   }
 
-  getRows(data) {
-    const rows = _.map(data, (row) => {
-      return _.omitBy(row, (v, k) => {
-        return COLS_BLACK_LIST.has(k);
+  getRows(data, cols) {
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: added comment

File: frontend/public/components/chargeback.tsx [hunk 16]
```diff
@@ -256,30 +277,37 @@ class ReportData extends React.Component<ReportDataProps, ReportDataState> {
...
+    let rows = _.map(data.results, values => {
+      return _.merge.apply(
+        {},
+        _.map(values.values, item => {
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: That is the yaml data I'm getting back from v2 api.

File: frontend/public/components/factory/table.tsx (hunk 1 above)
Comment (dtaylor113) [reply frontend/public/components/factory/table.tsx]: implemented 'sortAsNumber'.

File: frontend/public/style/_overrides.scss (hunk 2 above)
Comment (dtaylor113) [reply frontend/public/style/_overrides.scss]: yes, removed old code.

**Comment by dtaylor113** [comment]
Hi @spadgett, I believe I have addressed all of your review comments.  PTAL, thanks :-)

File: frontend/public/components/chargeback.tsx [hunk 17]
```diff
@@ -256,30 +270,82 @@ class ReportData extends React.Component<ReportDataProps, ReportDataState> {
...
+    //   ...
+    //  },
+    //  {
+    //   "period_start": "2019-01-01T00:00:00Z",
+    //   "pod_usage_cpu_core_seconds": 21.864000000000004
+    //   ...
+    //  },
+    //  ...
+     */
+    let rows = _.map(data.results, values => {
```
Comment (spadgett) [inline frontend/public/components/chargeback.tsx]:
```ts
const rowData = _.map(data.results, ({values}) => {
  return _.reduce(values, (acc, {name, value}) =>
    acc[name] = value;
    return acc;
  }, {});
});
```

But I suspect this can be simplified further  with the ordering below. Can you give an example of the final output after ordering by column? The code is hard to follow.

File: frontend/public/actions/ui.ts [hunk 18]
```diff
@@ -178,14 +178,14 @@ export const stopImpersonate = () => (dispatch) => {
   dispatch(detectFeatures());
   history.push(window.SERVER_FLAGS.basePath);
 };
-export const sortList = (listId: string, field: string, func: any, orderBy: string, column: string) => {
+export const sortList = (listId: string, field: string, func: any, sortAsNumber: boolean, orderBy: string, column: string) => {
```
Comment (spadgett) [inline frontend/public/actions/ui.ts]: Not new, but shouldn't `func` be `string` here?

This is actually another reason why we shouldn't pass a function as before. The value is stored in redux, and functions aren't serializable.

https://stackoverflow.com/questions/35325195/should-i-store-function-references-in-redux-store

File: frontend/public/components/factory/table.tsx [hunk 19]
```diff
@@ -403,15 +409,15 @@ export const Table = connect<TablePropsFromState,TablePropsFromDispatch,TablePro
       this._bodyRef = ref;
     }
 
-    _applySort(sortField, sortFunc, direction, columnTitle){
-      const {sortList, listId, currentSortFunc} = this.props;
+    _applySort(sortField, sortFunc, sortAsNumber, direction, columnTitle){
+      const {sortList, listId} = this.props;
       const applySort = _.partial(sortList, listId);
-      applySort(sortField, sortFunc || currentSortFunc, direction, columnTitle);
+      applySort(sortField, sortFunc, sortAsNumber, direction, columnTitle);
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/factory/table.tsx]: Patrick said it was legacy code from list.tsx.  It was messing up my implementation of `sortFunc` as a function in my last solution, I can restore it.

[the remark this answers is not in the document]
Comment (spadgett) [reply frontend/public/components/factory/table.tsx]: I would err on the side of keeping it unless we're certain we understand why it was there and why it's no longer needed.

[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/factory/table.tsx]: ok, I restored it.  The logic is still applicable if it's sortField OR sortFunc.

File: frontend/public/components/chargeback.tsx [hunk 20]
```diff
@@ -477,6 +543,17 @@ export type ReportDataState = {
 export type DataTableProps = {
   rows: any[],
   cols: string[],
+  schema: any,
```
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: fixed

File: frontend/public/components/chargeback.tsx (hunk 17 above)
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: > I don't see values: { values: ... } in the YAML. You're assigning the variable name on line 324.

The input is what is returned from the v2 Table api. `data.results.values[].value` etc..

I tried to implement the code you provided but it has errors, I'll work at getting past the them, but I'm a little concerned because my code does work...I'm not sure why we're looking for a different solution?

File: frontend/public/actions/ui.ts (hunk 18 above)
Comment (dtaylor113) [reply frontend/public/actions/ui.ts]: fixed

File: frontend/public/components/chargeback.tsx (hunk 17 above)
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: > Can you give an example of the final output after ordering by column? The code is hard to follow.

The code normalizes the order of columns/fields within each 'row':

image

Without the last bit of code/reordering, the reports look like this:
image

I filled: Reporting V2 Table API: 'columns' in results.values not in consistent order - https://bugzilla.redhat.com/show_bug.cgi?id=1736501 against 'Metering Operator' for this issue.

Comment (chancez) [reply frontend/public/components/chargeback.tsx]: PR to fix this is in-flight https://github.com/operator-framework/operator-metering/pull/895

Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: Rebased to latest `operator-framework/operator-metering`, verified @chancez fix.  Remove the column reordering logic/code.  Fixed

File: frontend/public/components/chargeback.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: This needs to be updated to:
```
<Timestamp noRelative={true} timestamp={value} />
```
Once, or after, https://github.com/openshift/console/pull/2341 is merged.

[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: This is fixed in https://github.com/openshift/console/pull/2341

**Comment by dtaylor113** [comment]
Hi @spadgett, I believe I have addressed all comments.   If you want, we can merge https://github.com/openshift/console/pull/2341 first, I can then update this PR to utilize https://github.com/openshift/console/pull/2341, then merge this PR...or...I can file a follow up PR.

File: frontend/public/components/chargeback.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (dtaylor113) [reply frontend/public/components/chargeback.tsx]: @chancez, I removed the `noRelative` logic from #2341 because that pr is a fix for our `Timestamp` component, regardless of chargeback report issues.  We need to decide if the fix in #2341 and/or the general `timestamp` displaying of dates is adequate for chargeback report results.

You wrote: "I think It may make sense [to show relative dates] in the details of the report, but i don't think it does in the report results."  

To that end, the fix in #2341 will only show relative dates +/- 10.5 minutes from 'now'.  I guess we need to decide if that is ok for dates in usage report results; particularly `period_start` and `period_end` dates.   If outside the +/- 10.5 minute range, then it will display the dates as "Mmm dd, yyyy t:ms a/p" , if date within current year it will display as "Mmm dd, t:ms a/p" -no year.  Is this ok for dates in report results?   If not, then in a follow up PR we can implement a `fullDate` property for `timestamp` component and update chargeback reports results -thanks

File: frontend/public/components/chargeback.tsx [hunk 21]
```diff
@@ -173,21 +184,24 @@ const DataCell = ({name, value}) => {
   return value;
 };
 
-const DataTable = ({cols, rows}:DataTableProps) => {
-  const size = _.clamp(Math.floor(12 / _.size(rows[0])), 1, 4);
-  const className = `col-md-${size}`;
+const DataTable = ({cols, rows, schema}: DataTableProps) => {
+
+  const getUnit = (col: string) => {
+    const colSchema = _.find(schema.values, {'name': col});
```
Comment (spadgett) [inline frontend/public/components/chargeback.tsx]:
```suggestion
    const colSchema = _.find(schema.values, {name: col});
```

File: frontend/public/components/chargeback.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (chancez) [reply frontend/public/components/chargeback.tsx]: I think we want to show the full date for results, always. Otherwise you cannot compare values across rows or cells easily.
